### ISSUE TITLE
TestingModule not calling constructor when class is Injectable and Scope.TRANSIENT

### ISSUE THREAD

**Issue by vschoener** [nestjs/nest#3419]
## Current behavior
With TestingModule, the injection is not made when the service uses the `Injectable` and `Scope.TRANSIENT`. Constructor seems not to be called (I put some log)

## Input Code
```ts
// Wrapper to implement Nest LoggerService Interface and use our custom logger
// src/core/logger/logger.service.ts
@Injectable({ scope: Scope.TRANSIENT })
export class LoggerService implements NestLoggerService {
  public prefix?: string;

  constructor(@Inject(LOGGER_PROVIDER) private logger: Logger) {}

  private formatMessage(message: string) {
    return this.prefix ? `[${this.prefix}] ${message}` : message;
  }

  private formatContext(context: unknown) {
    return context ? context : {};
  }

  setPrefix(prefix: string) {
    this.prefix = prefix;

    return this;
  }

  debug(message: string, context?: unknown): any {
    this.logger.debug(this.formatContext(context), this.formatMessage(message));
  }

  error(message: string, trace?: string, context?: unknown): any {
    this.logger.error(
      {
        context: this.formatContext(context),
        trace,
      },
      this.formatMessage(message),
    );
  }

  log(message: string, context?: unknown): any {
    this.logger.info(this.formatContext(context), this.formatMessage(message));
  }

  info(message: string, context?: unknown): any {
    this.logger.info(this.formatContext(context), this.formatMessage(message));
  }

  verbose(message: string, context?: unknown): any {
    this.logger.debug(this.formatContext(context), this.formatMessage(message));
  }

  warn(message: string, context?: unknown): any {
    this.logger.warn(this.formatContext(context), this.formatMessage(message));
  }
}
```

```ts
// src/core/logger/logger.provider.ts
import { FactoryProvider } from '@nestjs/common/interfaces';
import * as chprLogger from 'chpr-logger'; // Custom Logger I need to use
import { LOGGER_PROVIDER } from './constants';

export const loggerFactory: FactoryProvider = {
  provide: LOGGER_PROVIDER,
  useFactory: () => {
    return chprLogger.init({
      logger: {
        pretty: process.env.NODE_ENV !== 'production',
      },
    });
  },
};
```

```ts
// core/logger/logger.service.spec.ts

import { Test, TestingModule } from '@nestjs/testing';
import { Logger } from 'chpr-logger';
import { LoggerService } from './logger.service';
import { LOGGER_PROVIDER } from './constants';

describe('LoggerService', () => {
  let loggerService: LoggerService;
  let logger;

  beforeEach(async () => {
    const app: TestingModule = await Test.createTestingModule({
      providers: [
        {
          provide: LOGGER_PROVIDER,
          useValue: {
            debug: jest.fn(),
            warn: jest.fn(),
            info: jest.fn(),
            error: jest.fn(),
          },
        },
        LoggerService,
      ],
    }).compile();

    loggerService = await app.resolve<LoggerService>(LoggerService);
    logger = app.get<Logger>(LOGGER_PROVIDER); // This one is properly built
  });

  it('should forward log call', () => {
    // Throw an error saying can't call info on undefined because it has not been injected
    loggerService.log('Use log method');

    expect(logger.info.mock.calls).toEqual([[{}, 'Use log method']]);
  });
});
```

## Expected behavior
The logger should be injected in the LoggerService constructor

## Possible Solution
Actually I didn't look in the code to understand why

## Environment

<pre><code>
Nest version: 6.7.2
 
For Tooling issues:
- Node version: 12.13.0  
- Platform:  Mac

Others:
</code></pre>

**Comment by vschoener** [comment]
For your information, I find a way to unblock myself, I didn't put the Scope in the Injectable but I set this in the module like this, in this case, the TestingModule work as expected and I same when I run the process.
But it doesn't fix the main problem (or bad usage and understanding from me?)

```ts
import { Module, Scope } from '@nestjs/common';
import { loggerFactory } from './logger.provider';
import { LoggerService } from './logger.service';

@Module({
  providers: [
    loggerFactory,
    {
      provide: LoggerService,
      useClass: LoggerService,
      scope: Scope.TRANSIENT,
    },
  ],
  exports: [LoggerService],
})
export class LoggerModule {}
```

**Comment by kamilmysliwiec** [comment]
Please, provide a minimal repository which reproduces your issue.

**Comment by vschoener** [comment]
Yes, sorry for the delay I'll provide it soon

**Comment by vschoener** [comment]
@kamilmysliwiec here is the repo I just made  https://github.com/vschoener/nestjs-transient-test-issue
The last commit reproduces the issue, the previous one contains a way to not have the issue but I defined the Scope Transient in the providers from the Module.

If you need anything else, let me know

**Comment by jmcdo29** [comment]
@vschoener @kamilmysliwiec I'm having the same problem. Is there something special we have to do with `app.resolve`? When I check what my service value is I just get `ServiceClassName {}`. I started digging around with the module structure to see what I could find, but unfortunately the `WeakMap` that holds the value for the service just had the above under `instance`. I'll keep digging around to see if I can find out what's going on, just thought I'd say that I'm having the issue too :smile:

**Comment by vschoener** [comment]
@jmcdo29 Nice :) In my case, I did have the time to dig in the module but I'll try once I get the time :) That's really weird, maybe a promise issue not awaited somewhere?

**Comment by kamilmysliwiec** [comment]
Fixed in 6.10.4

**Comment by vschoener** [comment]
@kamilmysliwiec Awesome, good job!
